packages/api/src/external/ehr/elation/command/get-patients-from-appointments.ts

```diff
@@ -0,0 +1,249 @@
+    const { clientKey, clientSecret } = await getElationClientKeyAndSecret({
+      cxId,
+      practiceId,
+    });
+    const api = await ElationApi.create({
+      practiceId,
+      environment: elationEnvironment as ElationEnv,
+      clientKey,
+      clientSecret,
+    });
+    const appointments = await api.getAppointments({
+      cxId,
+      fromDate,
+      toDate,
+    });
+    patientAppointments.push(
+      ...appointments.map(appointment => {
+        return {
+          cxId,
+          elationPracticeId: practiceId,
+          elationPatientId: appointment.patient,
+        };
+      })
+    );
+  } catch (error) {
+    const cause = `Cause: ${errorToString(error)}`;
+    const details = `cxId ${cxId} practiceId ${practiceId}.`;
+    const msg = "Failed to get appointments.";
+    log(`${details} ${msg} ${cause}`);
+    errors.push(`${msg} ${details} ${cause}`);
+  }
+}
+
+async function getPatientIdOrFailByPractice({
+  cxId,
+  practiceId,
+  patientAppointmentsUnique,
+  errors,
+  log,
+}: {
+  cxId: string;
+  practiceId: string;
+  patientAppointmentsUnique: PatientAppointment[];
+  errors: string[];
+  log: typeof console.log;
+}) {
+  const { clientKey, clientSecret } = await getElationClientKeyAndSecret({
+    cxId,
+    practiceId,
+  });
+  const api = await ElationApi.create({
+    practiceId,
+    environment: elationEnvironment as ElationEnv,
+    clientKey,
+    clientSecret,
+  });
+  const getPatientIdOrFaiLArgs = patientAppointmentsUnique.map(appointment => {
+    return {
+      api,
+      cxId: appointment.cxId,
+      elationPracticeId: appointment.elationPracticeId,
+      elationPatientId: appointment.elationPatientId,
+      triggerDq: true,
+      errors,
+      log,
+    };
+  });
+
+  await executeAsynchronously(getPatientIdOrFaiLArgs, getPatientIdOrFail, {
+    numberOfParallelExecutions: parallelPatients,
+    delay: delayBetweenPracticeBatches.asMilliseconds(),
+  });
+}
+
+async function getPatientIdOrFail({
+  api,
+  cxId,
+  elationPracticeId,
+  elationPatientId,
+  triggerDq,
+  errors,
+  log,
+}: {
+  api: ElationApi;
+  cxId: string;
+  elationPracticeId: string;
+  elationPatientId: string;
+  triggerDq: boolean;
+  errors: string[];
+  log: typeof console.log;
+}): Promise<void> {
+  try {
+    await singleGetPatientIdOrFail({
+      api,
+      cxId,
+      elationPracticeId,
+      elationPatientId,
+      triggerDq,
+    });
+  } catch (error) {
+    const cause = `Cause: ${errorToString(error)}`;
+    const details = `cxId ${cxId} elationPracticeId ${elationPracticeId} elationPatientId ${elationPatientId}.`;
+    const msg = "Failed to find or create patients";
+    log(`${details} ${msg} ${cause}`);
+    errors.push(`${msg} ${details} ${cause}`);
+  }
+}
+
+function createMapKey(cxId: string, practiceId: string): string {
+  return `${cxId}|${practiceId}`;
+}
+
+function parseMapKey(key: string): { cxId: string; practiceId: string } {
+  const [cxId, practiceId] = key.split("|");
```

Comment:
100% sure `practiceId` can't have pipe on it?

Reply:
made it a triple pipe but this is elation and they actually get returned as numbers, so *very* unlikely it will ever contain a pipe.